pkg/controller/signer/signer-controller.go

```diff
@@ -0,0 +1,221 @@
+}
+
+// add adds a new Controller to mgr with r as the reconcile.Reconciler
+func add(mgr manager.Manager, r reconcile.Reconciler) error {
+	// Create a new controller
+	c, err := controller.New("signer-controller", mgr, controller.Options{Reconciler: r})
+	if err != nil {
+		return err
+	}
+
+	// Watch for changes to CetificateSigningRequest resource
+	err = c.Watch(&source.Kind{Type: &csrv1.CertificateSigningRequest{}}, &handler.EnqueueRequestForObject{})
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+var _ reconcile.Reconciler = &ReconcileCSR{}
+
+// ReconcileCSR reconciles a cluster CertificateSigningRequest object. This
+// will watch for changes to CertificateSigningRequest resources with
+// SignerName == signerName. It will automatically approve these requests for
+// signing. This assumes that the cluster has been configured in a way that
+// no bad actors can make certificate signing requests. In future, we may decide
+// to implement a scheme that would use a one-time token to validate a request.
+//
+// All requests will be signed using a CA, that is currently generated by
+// the OperatorPKI, and the signed certificate will be returned in the status.
+//
+// This allows clients to get a signed certificate while maintaining
+// private key confidentiality.
+type ReconcileCSR struct {
+	// This client, initialized using mgr.Client() above, is a split client
+	// that reads objects from the cache and writes to the apiserver
+	client client.Client
+	scheme *runtime.Scheme
+	status *statusmanager.StatusManager
+
+	// Note: We need a Clientset as the controller-runtime client does not
+	// support non-CRUD subresources (see
+	// https://github.com/kubernetes-sigs/controller-runtime/issues/452)
+	// This may risk invalidating the cache but in our case, this is not a
+	// problem as we only use this to update the approval status of the csr.
+	clientset *kubernetes.Clientset
+}
+
+// Reconcile CSR
+func (r *ReconcileCSR) Reconcile(request reconcile.Request) (reconcile.Result, error) {
+	csr := &csrv1.CertificateSigningRequest{}
+	err := r.client.Get(context.TODO(), request.NamespacedName, csr)
+	if err != nil {
+		if apierrors.IsNotFound(err) {
+			// Request object not found, could have been deleted after reconcile request.
+			// Return and don't requeue as the CSR has been deleted.
+			return reconcile.Result{}, nil
+		}
+		// Error reading the object - requeue the request.
+		log.Println(err)
+
+		return reconcile.Result{}, err
+	}
+
+	// Only handle CSRs for this signer
+	if csr.Spec.SignerName != signerName {
+		// Don't handle a CSR for another signerName. We don't need to log this as
+		// we will pollute the logs. We also don't need to requeue it.
+		return reconcile.Result{}, nil
+	}
+
+	if isCertificateRequestApproved(csr) {
+		// We will not recertify an approved certificate
+		return reconcile.Result{}, nil
```

Comment:
What happens if we approve a CSR then crash?

Reply:
I updated this code. Because the watch doesn't have predicates, all updates, creates and deletes get handled by the reconciler. As I update the CSR twice in the flow (once for adding the approval and again for adding the certificate), there needs to be a few guards there to not process the updates.

Reply:
Multiple reconciliations are fine, as long as you're level-triggered.

You need to handle objects in all possible states, so if a certificate is approved, then we move on and sign it.

Reply:
Updated, if we approve and crash, it will attempt to resign.